### PR TITLE
The intuition behind rusts borrowing rules and ownership

### PR DESCRIPTION
**Pull request by RainingComputers**

### CODE REVIEW

File: the_intuition_behind_rusts_borrowing_rules_and_ownership.md [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+
+The concept of ownership and borrowing was not a very obvious one for me when I first looked at it. After spending some time with Rust and a lot of thinking it started to become clear.
+
+In this blog I will try to capture the thoughts that finally made this concept click for me in the hope that it might help others also. I will go over my thoughts on why the borrowing in Rust works the way it does and my interpretation of the borrowing rules. You can copy and run the examples at the [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021) as you follow the blog.
+
+## What is the borrow checker trying to do?
+
+Let's first take a look at some Python code that demonstrates a common bug that happens in reference based languages
+
+```python {.numberLines}
```
Comment (alexsanjoseph) [inline the_intuition_behind_rusts_borrowing_rules_and_ownership.md:15]: TIL about `.numberLines`. Thanks

File: the_intuition_behind_rusts_borrowing_rules_and_ownership.md [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+The error points out that
+
+- `b` was pointing to an existing value `a` at line 4
+- An unexpected modification happened behind `b`'s back at line 5
+- That unexpected modification is seen or "used" at line 7
+
+The borrow checker prevented us from using `b` at line 7 because the value it was pointing to was unexpectedly modified behind its back at line 5.
+
+Let's try to make this error go away. One way is to make the unexpected modification an expected one. You can do that by changing `a.push(4)` to `b.push(4)`. The line that does the modification now has the letter `b` in it and the modification is not done behind `b`'s back. Make the change and run the program, you can see it compiles successfully and prints `[1, 2, 3, 4]`!
+
```
Comment (alexsanjoseph) [inline the_intuition_behind_rusts_borrowing_rules_and_ownership.md:70]: Natural question was what happens if I do print(a) after this. That also becomes [1,2,3,4], which was not exactly intuitive? Might make sense as an addendum

File: the_intuition_behind_rusts_borrowing_rules_and_ownership.md [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+    // compiler automatically inserts code to call drop
+    //  because a goes out of scope here
+}
+```
+
+This program will print `Dropping Foo with x 100!`. We create a `struct` (or `class`) called foo, then we add a method[^3] called `drop` to it. The `drop` method is a special one, the compiler automatically inserts code to call this method on owners when they go out of scope. The `Vec` types (created by calling `vec!`) in Rust use this method to free memory that was allocated for the items.
+
+## Conclusion
+
+Hopefully this blog post has captured some of the reasons for why Rust's borrowing and ownership works the way it does and how it is different from other languages. This blog post is not meant to be complete resource on learning borrowing and ownership, so I highly recommend reading the Rust book and other resources. I had originally planned to write about lifetimes also, but this post has already become a long one, so stay tuned for another post on lifetimes!
```
Comment (alexsanjoseph) [inline the_intuition_behind_rusts_borrowing_rules_and_ownership.md:418]: W.r.t lifetimes a nice idea would be to ask a question on why `lifetimes` will be relevant? 

```
But what about the case when A goes to B and then becomes C? To understand that you need to understand lifetimes. I had originally...
```

Then you can choose to drop the teaser for next post.

File: the_intuition_behind_rusts_borrowing_rules_and_ownership.md (hunk 1 above)
Comment (RainingComputers) [reply the_intuition_behind_rusts_borrowing_rules_and_ownership.md:15]: Yeah, this one took some time to find out

File: the_intuition_behind_rusts_borrowing_rules_and_ownership.md (hunk 3 above)
Comment (RainingComputers) [reply the_intuition_behind_rusts_borrowing_rules_and_ownership.md:418]: Yes

File: the_intuition_behind_rusts_borrowing_rules_and_ownership.md (hunk 2 above)
Comment (RainingComputers) [reply the_intuition_behind_rusts_borrowing_rules_and_ownership.md:70]: Hmm, will look into this one

**Pull request merged**